File: September/28/removedupesorted2.py
# LeetCode 82: Remove Duplicates from Sorted List II
# Explanation:
# 1. Use dummy node to handle edge cases (duplicates at head).
# 2. Iterate with prev and current pointers.
# 3. Skip nodes while duplicates exist.
# 4. Attach prev.next to next distinct node.
# Time Complexity: O(n)
# Space Complexity: O(1)

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def deleteDuplicates(self, head: ListNode) -> ListNode:
        dummy = ListNode(0, head)
        prev = dummy
        while head:
            if head.next and head.val == head.next.val:
                while head.next and head.val == head.next.val:
                    head = head.next
                prev.next = head.next
            else:
                prev = prev.next
            head = head.nex
        return dummy.next
    
    deleteDuplicatesOneLine=lambda s,h:None if not h else h if not h.next or h.val!=h.next.val else s.deleteDuplicates(next((n for n in iter(lambda:h: (h:=h.next) and h, None),None)))

File: 2025/September/28/removedupesorted2.java
// LeetCode 82: Remove Duplicates from Sorted List II
// Explanation:
// 1. Use dummy node to simplify head removal.
// 2. Skip all nodes with duplicate values.
// 3. Connect previous pointer to next distinct node.
// Time Complexity: O(n)
// Space Complexity: O(1)

class ListNode {
    int val;
    ListNode next;
    ListNode() {}
    ListNode(int val) { this.val = val; }
    ListNode(int val, ListNode next) { this.val = val; this.next = next; }
}

class Solution {
    public ListNode deleteDuplicates(ListNode head) {
        ListNode dummy = new ListNode(0, head);
        ListNode prev = dummy;
        while (head != null) {
            if (head.next != null && head.val == head.next.val) {
                while (head.next != null && head.val == head.next.val) {
                    head = head.next;
                }
                prev.next = head.next;
            } else {
                prev = prev.next;
            }
            head = head.next;
        }
        return dummy.next;
    }
}

// Example usage:
// public static void main(String[] args) {
//     Solution sol = new Solution();
//     ListNode head = new ListNode(1, new ListNode(2, new ListNode(3,
//                     new ListNode(3, new ListNode(4, new ListNode(4, new ListNode(5)))))));
//     ListNode result = sol.deleteDuplicates(head);
//     while (result != null) {
//         System.out.print(result.val + " "); // 1 2 5
//         result = result.next;
//     }
// }
